### ISSUE TITLE
deftype, defrecord, and constructor calls?

### ISSUE THREAD

**Issue by sogaiu** [carocad/parcera#83]
It looks like there is currently no specific support for deftype, defrecord, or constructor calls as described here: https://clojure.org/reference/reader#_deftype_defrecord_and_constructor_calls_version_1_3_and_later

I get this from `ast`:
```
parcera.core=> (ast "#my.klass_or_type_or_record[:a :b :c]")
(:code (:tag (:symbol "my.klass_or_type_or_record") (:vector (:keyword ":a") (:whitespace " ") (:keyword ":b") (:whitespace " ") (:keyword ":c"))))
```
and:
```
parcera.core=> (ast "#my.record{:a 1, :b 2}")
(:code (:tag (:symbol "my.record") (:map (:keyword ":a") (:whitespace " ") (:number "1") (:whitespace ", ") (:keyword ":b") (:whitespace " ") (:number "2"))))
```

So IIUC it looks like these are currently recognized as tagged literals (`failure?` returned nil for both FWIW).

Here is a brief transcript of a repl session for the record case:
```
user=> (defrecord Fun [a b])
user.Fun
user=> #user.Fun[1 2]
#user.Fun{:a 1, :b 2}
user=> #user.Fun{:a 1 :b 2}
#user.Fun{:a 1, :b 2}
```

I think some relevant lines in LispReader may be:

https://github.com/clojure/clojure/blob/0035cd8d73517e7475cb8b96c7911eb0c43a1a9d/src/jvm/clojure/lang/LispReader.java#L1451-L1502

**Comment by carocad** [comment]
@sogaiu in my opinion parcera is doing the right thing here. In terms of grammar, there is no clear distinction between a tagged literal and a literal constructor call. You can also see this in the LispReader [here](https://github.com/clojure/clojure/blob/0035cd8d73517e7475cb8b96c7911eb0c43a1a9d/src/jvm/clojure/lang/LispReader.java#L1427) where the `.` character is used to distinguish between the two, however on both cases a [symbol](https://github.com/clojure/clojure/blob/0035cd8d73517e7475cb8b96c7911eb0c43a1a9d/src/jvm/clojure/lang/LispReader.java#L1419) is expected.

This is a clear example of runtime behavior been different than reader behavior. In Clojure, both steps are performed together as you can see in the LispReader checking the state of certain maps/atoms and acting accordingly.

Closing this as I don't see this as an issue.

**Comment by sogaiu** [comment]
Thanks for the explanation including the reference to the source code.

The point about no clear distinction between a tagged literal and a literal constructor call seems right to me too (in terms of the grammar).

On a side note, looking at the code for `CtorReader` (https://github.com/carocad/parcera/blob/1c513564fa6549fef7d39477b8bc359380f1942e/src/clojure/parcera/core.cljc#L69-L70) I am noticing that `readTagged` and `readRecord` are methods within the `CtorReader` class.

That got me wondering a bit about the naming of the construct within the grammar.  IIUC, it's currently `tag`, but I am wondering whether `ctor` might be another choice.

**Comment by carocad** [comment]
I am only guessing here but I think that it is called `CtorReader` just out of lack of better names. 

In general, the LispReader has a dispatch mechanism by checking individual characters for Clojure the `#` is the [dispatch reader](https://github.com/clojure/clojure/blob/06097b1369c502090be6489be27cc280633cb1bd/src/jvm/clojure/lang/LispReader.java#L106) not a tagged literal as explained in most references. Inside the dispatch reader, it will try to find a macro first and default to the CtorReader otherwise; which as we discussed before is both a Constructor and a Tag literal dispatcher.

Find a proper name for such a mix of functionality is pretty hard, both at the grammar level and at the runtime level. Hope it helps

**Comment by sogaiu** [comment]
Thanks for sharing your thoughts, I appreciate it as I am making choices for names elsewhere.

To be clear, the following is just for the sake of understanding and investigation.

I did a bit of searching and it looks to me like tagged reader literals (readTagged) were added after the mechanism for "deftype, defrecord, and constructor calls" (readRecord): https://github.com/clojure/clojure/commit/d3b5665d21457ad27bda702f567ca2f55b14283b

It seems that `CtorReader` was already in place and it happens that `readTagged` would fit right in alongside `readRecord`.  (Perhaps this is less of a coincidence -- may be this was about the only place it could fit.)

It reminded me a bit of how set is not grouped along with list, map, and vector (as a collection in the grammar) because set alone is handled via dispatch (appreciated the comment in the source btw).  Although in the present case, it's the reverse -- that due to grammatical structure, there are two things that are being grouped together which perhaps aren't really all that related apart from surface structure.

My impression is that tagged literals are more well-known than the other construct though, so in terms of familiarity I could see a case for using `tag` as a name for both types in the grammar.
